### ISSUE TITLE
Example request for training a network with multiple inputs with different loaders

### ISSUE THREAD

**Issue by achbogga** [pytorch/xla#895]
I have gone through the API provided for XLA and also the examples from the xla/test. I am trying a train a custom two-stage network which has two different CNN backbones, which takes in two different inputs from two different data loaders. How do you go about writing code for such networks? An official example would be of great help.

Thanks a lot in advance.

**Comment by taylanbil** [comment]
So, is your input at every step a pair of tensors? if so, you should be able to `zip` those two loaders and get one loader, and change `train_loop_fn` appropriately, as in

```python
for step, (tensor1, tensor2) in loader:
  ...
```
Does that work?

**Comment by achbogga** [comment]
@taylanbil The lengths of the loaders (tensor dim 0 or batch size) are different, to zip them into tuples.

**Comment by taylanbil** [comment]
I am not sure if I understand; so you have `loader1`, and `loader2`.

Is the batch size in `loader1` always the same at every step? If not, you'll get bad performance as TPUs like static shapes

Is there some pattern in which you consume tensors from different loaders? For example, do you do 3 steps from loader1 and 2 steps from loader2? If so, that should be solvable with a zip-like function too. you can then custom prepare your loader, and do
```python
for step, ((x1, x2, x3), (y1, y2)) in loader:
 ...
```

**Comment by dlibenzi** [comment]
> I have gone through the API provided for XLA and also the examples from the xla/test. I am trying a train a custom two-stage network which has two different CNN backbones, which takes in two different inputs from two different data loaders. How do you go about writing code for such networks? An official example would be of great help.
> 
> Thanks a lot in advance.

Can you create a very small snippet about your code, so we can get an idea of what is really happening?

**Comment by achbogga** [comment]
- To give you guys a little bit more background. I am working with code forked from [this repo](https://github.com/floodsung/LearningToCompare_FSL) to train Relation Networks for few-shot learning scenario. 
- Initially, this training process consists of two different CNNs, but I have combined them into a single network which can accept two inputs given via two different loaders. However, I have a file called task_generator_tpu.py, included below, which is called twice to get two different types of loaders with two different number of samples. 
- The calls are also given below. Now, I need to create a single loader which can give me both types of input tensors.

### The task_generator_tpu.py for generating two different loaders

```
import torchvision
import torchvision.datasets as dset
import torchvision.transforms as transforms
import torch
from torch.utils.data import DataLoader,Dataset
import random
import os
from PIL import Image
import matplotlib.pyplot as plt
import numpy as np
from torch.utils.data.sampler import Sampler

def imshow(img):
	npimg = img.numpy()
	plt.axis("off")
	plt.imshow(np.transpose(npimg,(1,2,0)))
	plt.show()

class Rotate(object):
	def __init__(self, angle):
		self.angle = angle
	def __call__(self, x, mode="reflect"):
		x = x.rotate(self.angle)
		return x

def china_drinks_sku_folders(data_folder = '/home/caffe/achu/Data/china_drinks/image_data/cropped_images', no_of_training_samples = 10, no_of_validation_samples = 20, validation_split_percentage = 0.1):

	sku_folders = [os.path.join(data_folder, family, sku) \
				for family in os.listdir(data_folder) \
				if os.path.isdir(os.path.join(data_folder, family)) \
				for sku in os.listdir(os.path.join(data_folder, family))]
	random.seed(1)
	random.shuffle(sku_folders)

	print ('Total SKUS in the dataset: ', len(sku_folders))
	req_samples = no_of_training_samples+no_of_validation_samples
	sku_folders = [folder for folder in sku_folders if len(os.listdir(folder))>=req_samples]
	print ('Total SKUS in the dataset with '+str(req_samples)+' samples: ', len(sku_folders))

	val_split = int(validation_split_percentage*len(sku_folders))
	train_split = len(sku_folders) - val_split

	metatrain_sku_folders = sku_folders[:train_split]
	print ('Train split SKUS: ', len(metatrain_sku_folders))
	#print ('Debug: ', metatrain_sku_folders[0])
	metaval_sku_folders = sku_folders[train_split:]
	print ('Val split SKUS: ', len(metaval_sku_folders))

	return metatrain_sku_folders,metaval_sku_folders

class ChinaDrinksTask(object):
	# This class is for task generation for both meta training and meta testing.
	# For meta training, we use all 20 samples without valid set (empty here).
	# For meta testing, we use 1 or 5 shot samples for training, while using the same number of samples for validation.
	# If set num_samples = 20 and chracter_folders = metatrain_sku_folders, we generate tasks for meta training
	# If set num_samples = 1 or 5 and chracter_folders = metatest_chracter_folders, we generate tasks for meta testing
	def __init__(self, sku_folders, num_classes, train_num,test_num):

		self.sku_folders = sku_folders
		self.num_classes = num_classes
		self.train_num = train_num
		self.test_num = test_num

		#print ('Sampling :', num_classes, ' from ', len(sku_folders))

		class_folders = random.sample(self.sku_folders,self.num_classes)
		selected_skus = [folder.split('/')[-1] for folder in class_folders]
		labels_ar = np.array(range(len(selected_skus)))
		labels = dict(zip(selected_skus, labels_ar))
		samples = dict()

		self.train_roots = []
		self.test_roots = []
		for c in class_folders:

			temp = [os.path.join(c, x) for x in os.listdir(c)]
			samples[c] = temp
			#print ('Debug0: ', len(temp))

			self.train_roots += samples[c][:train_num]
			self.test_roots += samples[c][train_num:(train_num+test_num)]
		#print ('Debug: ', self.train_roots[0])

		self.train_labels = [labels[self.get_class(x)] for x in self.train_roots]
		self.test_labels = [labels[self.get_class(x)] for x in self.test_roots]

	def get_class(self, sample):
		#print ('Debug: ', sample)
		return str(sample.split('/')[-2])


class FewShotDataset(Dataset):

	def __init__(self, task, image_size = 160, split='train', transform=None, target_transform=None):
		self.transform = transform # Torch operations on the input image
		self.target_transform = target_transform
		self.task = task
		self.split = split
		self.image_size = image_size
		self.image_roots = self.task.train_roots if self.split == 'train' else self.task.test_roots
		self.labels = self.task.train_labels if self.split == 'train' else self.task.test_labels

	def __len__(self):
		return len(self.image_roots)

	def __getitem__(self, idx):
		raise NotImplementedError("This is an abstract class. Subclass this class for your particular dataset.")


class ChinaDrinks(FewShotDataset):

	def __init__(self, *args, **kwargs):
		super(ChinaDrinks, self).__init__(*args, **kwargs)

	def __getitem__(self, idx):
		try:
			image_root = self.image_roots[idx]
			image = Image.open(image_root)
			image = image.convert('L')
			image = image.resize((self.image_size,self.image_size), resample=Image.LANCZOS) # per Chelsea's implementation
			#image = np.array(image, dtype=np.float32)
			if self.transform is not None:
				image = self.transform(image)
			label = self.labels[idx]
			if self.target_transform is not None:
				label = self.target_transform(label)
			return image, label
		except IndexError:
			print ('Index Error: ', idx, len(self.image_roots))
			return None, None

class ClassBalancedSampler(Sampler):
	''' Samples 'num_inst' examples each from 'num_cl' pools
		of examples of size 'num_per_class' '''

	def __init__(self, num_per_class, num_cl, num_inst,shuffle=True):
		self.num_per_class = num_per_class
		self.num_cl = num_cl
		self.num_inst = num_inst
		self.shuffle = shuffle

	def __iter__(self):
		# return a single list of indices, assuming that items will be grouped by class
		if self.shuffle:
			batch = [[i+j*self.num_inst for i in torch.randperm(self.num_inst)[:self.num_per_class]] for j in range(self.num_cl)]
		else:
			batch = [[i+j*self.num_inst for i in range(self.num_inst)[:self.num_per_class]] for j in range(self.num_cl)]
		batch = [item for sublist in batch for item in sublist]

		if self.shuffle:
			random.shuffle(batch)
		return iter(batch)

	def __len__(self):
		return 1


def get_data_loader(task, image_size = 160, num_per_class=1, split='train',shuffle=True,rotation=0, num_workers = 3):
	# NOTE: batch size here is # instances PER CLASS
	normalize = transforms.Normalize(mean=[0.92206, 0.92206, 0.92206], std=[0.08426, 0.08426, 0.08426])

	dataset = ChinaDrinks(task, image_size = image_size, split=split, transform=transforms.Compose([Rotate(rotation),transforms.ToTensor()]))#,normalize]))

	if split == 'train':
		sampler = ClassBalancedSampler(num_per_class, task.num_classes, task.train_num,shuffle=shuffle)
	else:
		sampler = ClassBalancedSampler(num_per_class, task.num_classes, task.test_num,shuffle=shuffle)
	loader = DataLoader(dataset, batch_size=num_per_class*task.num_classes, sampler=sampler)

	return loader

 ```

### Training snippet is as given below

```
for episode in range(EPISODE):

        relation_network_scheduler.step(episode)

        # init dataset
        # sample_dataloader is to obtain previous samples for compare
        # batch_dataloader is to batch samples for training
        degrees = random.choice([0,90,180,270])
        task = tgtpu.ChinaDrinksTask(metatrain_character_folders,CLASS_NUM,SAMPLE_NUM_PER_CLASS,QUERY_NUM_PER_CLASS)
        sample_dataloader = tgtpu.get_data_loader(task, image_size = IMAGE_SIZE, num_per_class=SAMPLE_NUM_PER_CLASS,split="train",shuffle=False,rotation=degrees)
        batch_dataloader = tgtpu.get_data_loader(task, image_size = IMAGE_SIZE, num_per_class=QUERY_NUM_PER_CLASS,split="test",shuffle=True,rotation=degrees)


        # sample datas
        samples,sample_labels = sample_dataloader.__iter__().next()
        batches,batch_labels = batch_dataloader.__iter__().next()
        
        relation_scores = relation_network(Variable(samples).cuda(GPU), Variable(batches).cuda(GPU))
        relations = relation_scores.view(-1,CLASS_NUM)

        mse = nn.MSELoss().cuda(GPU)
        one_hot_labels = Variable(torch.zeros(QUERY_NUM_PER_CLASS*CLASS_NUM, CLASS_NUM).scatter_(1, batch_labels.view(-1,1), 1)).cuda(GPU)
        loss = mse(relations,one_hot_labels)


        # training

        relation_network.zero_grad()
        loss.backward()
        torch.nn.utils.clip_grad_norm_(relation_network.parameters(),0.5)
        relation_network_optim.step()
```

- I need to somehow get the samples and batches tensors from a single loader file now. Any help would be appreciated.
- I know that the above training calls are not in a train_fn_loop fashion which is required for TPU training. But that would be trivial if we have a single loader I guess.

**Comment by dlibenzi** [comment]
> @taylanbil The lengths of the loaders (tensor dim 0 or batch size) are different, to zip them into tuples.

It should not matter AFAICT.
Why can't you have a loader which returns something like?
```
return (samples, sample_labels, batches,batch_labels)
```

There will be one tuple per step, which you then unpack.

**Comment by achbogga** [comment]
@dlibenzi, I got this loader to work in this fashion. But I had to make a lot of changes to my code. But now, I am facing other issues. I will try to debug them and if everything works, I will share my code. Thanks, everyone.

**Comment by dlibenzi** [comment]
Closing this, but please reopen if you have more info.
